File: frontend/types/activity.ts
export enum ActivityType {
  TASK = 'TASK',
  EMAIL = 'EMAIL',
  CALL = 'CALL',
  MEETING = 'MEETING',
  NOTE = 'NOTE',
}

export enum ActivityStatus {
  PENDING = 'PENDING',
  IN_PROGRESS = 'IN_PROGRESS',
  COMPLETED = 'COMPLETED',
  CANCELLED = 'CANCELLED',
}

export enum ActivityPriority {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  URGENT = 'URGENT',
}

export interface Activity {
  id: string;
  activityId: string;

  // Basic Information
  subject: string;
  type: ActivityType;
  status: ActivityStatus;
  priority: ActivityPriority;
  description?: string;

  // Scheduling
  scheduledDate?: string;
  dueDate?: string;
  completedDate?: string;
  durationMinutes?: number;
  location?: string;

  // Related Entities
  leadId?: string;
  leadName?: string;
  contactId?: string;
  contactName?: string;
  accountId?: string;
  accountName?: string;
  opportunityId?: string;
  opportunityName?: string;

  // Assignment
  assignedToId?: string;
  assignedToName?: string;
  participants?: string[];

  // Email Specific
  emailFrom?: string;
  emailTo?: string;
  emailCc?: string[];
  emailBcc?: string[];
  emailSubject?: string;

  // Call Specific
  phoneNumber?: string;
  callDirection?: string;
  callOutcome?: string;
  callDuration?: number;

  // Meeting Specific
  meetingLink?: string;
  meetingType?: string;
  attendees?: string[];

  // Task Specific
  taskCategory?: string;
  isRecurring?: boolean;
  recurrencePattern?: string;

  // Additional Information
  tags?: string[];
  outcome?: string;
  nextSteps?: string;
  isPrivate?: boolean;
  reminderSet?: boolean;
  reminderDate?: string;

  // System Fields
  createdAt: string;
  createdBy: string;
  createdByName: string;
  lastModifiedAt: string;
  lastModifiedBy: string;
  lastModifiedByName: string;
}

export interface CreateActivityRequest {
  subject: string;
  type: ActivityType;
  status: ActivityStatus;
  priority?: ActivityPriority;
  description?: string;

  // Scheduling
  scheduledDate?: string;
  dueDate?: string;
  durationMinutes?: number;
  location?: string;

  // Related Entities
  leadId?: string;
  contactId?: string;
  accountId?: string;
  opportunityId?: string;

  // Assignment
  assignedToId?: string;
  participants?: string[];

  // Email Specific
  emailFrom?: string;
  emailTo?: string;
  emailCc?: string[];
  emailBcc?: string[];
  emailSubject?: string;

  // Call Specific
  phoneNumber?: string;
  callDirection?: string;
  callOutcome?: string;
  callDuration?: number;

  // Meeting Specific
  meetingLink?: string;
  meetingType?: string;
  attendees?: string[];

  // Task Specific
  taskCategory?: string;
  isRecurring?: boolean;
  recurrencePattern?: string;

  // Additional
  tags?: string[];
  outcome?: string;
  nextSteps?: string;
  isPrivate?: boolean;
  reminderSet?: boolean;
  reminderDate?: string;
}

export interface UpdateActivityRequest {
  subject?: string;
  type?: ActivityType;
  status?: ActivityStatus;
  priority?: ActivityPriority;
  description?: string;

  // Scheduling
  scheduledDate?: string;
  dueDate?: string;
  completedDate?: string;
  durationMinutes?: number;
  location?: string;

  // Related Entities
  leadId?: string;
  contactId?: string;
  accountId?: string;
  opportunityId?: string;

  // Assignment
  assignedToId?: string;
  participants?: string[];

  // Email Specific
  emailFrom?: string;
  emailTo?: string;
  emailCc?: string[];
  emailBcc?: string[];
  emailSubject?: string;

  // Call Specific
  phoneNumber?: string;
  callDirection?: string;
  callOutcome?: string;
  callDuration?: number;

  // Meeting Specific
  meetingLink?: string;
  meetingType?: string;
  attendees?: string[];

  // Task Specific
  taskCategory?: string;
  isRecurring?: boolean;
  recurrencePattern?: string;

  // Additional
  tags?: string[];
  outcome?: string;
  nextSteps?: string;
  isPrivate?: boolean;
  reminderSet?: boolean;
  reminderDate?: string;
}

export interface ActivityStatistics {
  // Total counts
  totalActivities: number;
  activeActivities: number;
  completedActivities: number;
  cancelledActivities: number;
  overdueActivities: number;

  // By type
  taskCount: number;
  emailCount: number;
  callCount: number;
  meetingCount: number;
  noteCount: number;

  // By status
  pendingCount: number;
  inProgressCount: number;
  completedCount: number;

  // By priority
  urgentCount: number;
  highCount: number;
  mediumCount: number;
  lowCount: number;

  // Time metrics
  averageDurationMinutes: number;
  totalCallDuration: number;
  totalMeetingDuration: number;
}

File: backend/src/main/java/com/ultron/backend/domain/entity/Activity.java
package com.ultron.backend.domain.entity;

import com.ultron.backend.domain.enums.ActivityType;
import com.ultron.backend.domain.enums.ActivityStatus;
import com.ultron.backend.domain.enums.ActivityPriority;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

@Document(collection = "activities")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Activity {
    @Id
    private String id;

    @Indexed(unique = true)
    private String activityId; // ACT-YYYY-MM-XXXXX

    // Multi-tenancy
    @Indexed
    private String tenantId;

    // Basic Information
    private String subject;
    private ActivityType type;
    private ActivityStatus status;
    private ActivityPriority priority;
    private String description;

    // Scheduling
    private LocalDateTime scheduledDate;
    private LocalDateTime dueDate;
    private LocalDateTime completedDate;
    private Integer durationMinutes;
    private String location;

    // Related Entities
    private String leadId;
    private String leadName;
    private String contactId;
    private String contactName;
    private String accountId;
    private String accountName;
    private String opportunityId;
    private String opportunityName;

    // Assignment
    private String assignedToId;
    private String assignedToName;
    private List<String> participants;

    // Email Specific
    private String emailFrom;
    private String emailTo;
    private List<String> emailCc;
    private List<String> emailBcc;
    private String emailSubject;

    // Call Specific
    private String phoneNumber;
    private String callDirection; // INBOUND, OUTBOUND
    private String callOutcome;
    private Integer callDuration;

    // Meeting Specific
    private String meetingLink;
    private String meetingType; // IN_PERSON, VIRTUAL, PHONE
    private List<String> attendees;

    // Task Specific
    private String taskCategory;
    private Boolean isRecurring;
    private String recurrencePattern;

    // Additional Information
    private List<String> tags;
    private String outcome;
    private String nextSteps;
    private Boolean isPrivate;
    private Boolean reminderSet;
    private LocalDateTime reminderDate;

    // System Fields
    private LocalDateTime createdAt;
    private String createdBy;
    private String createdByName;
    private LocalDateTime lastModifiedAt;
    private String lastModifiedBy;
    private String lastModifiedByName;
    private Boolean isDeleted;
    private String deletedBy;
    private LocalDateTime deletedAt;
}
